src/impl_owned_array.rs

```diff
@@ -59,3 +69,558 @@ where
+                debug_assert!(tail_view.is_standard_layout(),
+                              "not std layout dim: {:?}, strides: {:?}",
+                              tail_view.shape(), tail_view.strides());
+            } 
+
+            // Keep track of currently filled lenght of `self.data` and update it
+            // on scope exit (panic or loop finish).
+            struct SetLenOnDrop<'a, A: 'a> {
+                len: usize,
+                data: &'a mut OwnedRepr<A>,
+            }
+
+            impl<A> Drop for SetLenOnDrop<'_, A> {
+                fn drop(&mut self) {
+                    unsafe {
+                        self.data.set_len(self.len);
+                    }
+                }
+            }
+
+            let mut data_length_guard = SetLenOnDrop {
+                len: self.data.len(),
+                data: &mut self.data,
+            };
+
+
+            // Safety: tail_view is constructed to have the same shape as array
+            Zip::from(tail_view)
+                .and_unchecked(array)
+                .debug_assert_c_order()
+                .for_each(|to, from| {
+                    to.write(from.clone());
+                    data_length_guard.len += 1;
+                });
+            drop(data_length_guard);
+
+            // update array dimension
+            self.strides = strides;
+            self.dim = res_dim;
+        }
+        // multiple assertions after pointer & dimension update
+        debug_assert_eq!(self.data.len(), self.len());
+        debug_assert_eq!(self.len(), new_len);
+        debug_assert!(self.pointer_is_inbounds());
+
+        Ok(())
+    }
+}
+
+/// Sort axes to standard order, i.e Axis(0) has biggest stride and Axis(n - 1) least stride
+///
+/// The axes should have stride >= 0 before calling this method.
+fn sort_axes_in_default_order<S, D>(a: &mut ArrayBase<S, D>)
+where
+    S: RawData,
+    D: Dimension,
+{
+    if a.ndim() <= 1 {
+        return;
+    }
+    sort_axes1_impl(&mut a.dim, &mut a.strides);
+}
+
+/// Sort axes to standard order, i.e Axis(0) has biggest stride and Axis(n - 1) least stride
+///
+/// The axes should have stride >= 0 before calling this method.
+fn sort_axes_in_default_order_tandem<S, S2, D>(a: &mut ArrayBase<S, D>, b: &mut ArrayBase<S2, D>)
+where
+    S: RawData,
+    S2: RawData,
+    D: Dimension,
+{
+    if a.ndim() <= 1 {
+        return;
+    }
+    sort_axes_impl(&mut a.dim, &mut a.strides, &mut b.dim, &mut b.strides);
+}
+
+fn sort_axes1_impl<D>(adim: &mut D, astrides: &mut D)
+where
+    D: Dimension,
+{
+    debug_assert!(adim.ndim() > 1);
+    debug_assert_eq!(adim.ndim(), astrides.ndim());
+    // bubble sort axes
+    let mut changed = true;
+    while changed {
+        changed = false;
+        for i in 0..adim.ndim() - 1 {
+            let axis_i = i;
+            let next_axis = i + 1;
+
+            // make sure higher stride axes sort before.
```

Comment:
Fwiw, I'd consider sorting length 1 axes to be the outermost loops, although I doubt it would have much effect either way.

Reply:
That's interesting, and also another uncommon but low hanging fruit would be removing dim 1 axes for dyn dim.

Reply:
I actually have another WIP on this topic in Baseiter - low hanging fruit is to swap axis `n - 2` and `n - 1`  if `Axis(n - 1)` is length 1.

Reply:
> I actually have another WIP on this topic in Baseiter - low hanging fruit is to swap axis `n - 2` and `n - 1` if `Axis(n - 1)` is length 1.

A somewhat more general improvement is to merge axes, when possible, into axis `n-1`, with something like this:

```rust
let mut take = n-2;
loop {
    let merged = arr.merge_axes(Axis(take), Axis(n-1));
    if !merged || take == 0 {
        break;
    } else {
        take -= 1;
    }
}
```

For example, in addition to cases where axis `n-1` has length 1, this helps cases where a standard layout array has been sliced with a step larger than 1 for the outermost axis.

> With that said, the goal of the axis sorting here is iteration in memory order, not performance.

Right, it probably doesn't matter much here.

Reply:
I think I managed to write the same as that in #979

Reply:
I like your implementation better. :)